design/build-clusters.md

```diff
@@ -0,0 +1,163 @@
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```
+
+
+### New table for build cluster details
+
+Table: `buildCluster`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | yes | no | no | |
+| `scmContext` | text(200) | no | yes | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
+| `managedByEmail` | text(100) | yes | no | no | cluster admin email for communications |
+
+Unique constraint: `name + isActive` 
+
+#### Sample record
+
+| id | name | scmContext | scmOrganizations | isActive | managedBy | managedByEmail 
+| --- | --- | --- | --- | --- | --- | --- | 
+| 1 | gq1 | github:git.ouroath.com | null | 0 | screwdriver | sd@oath.com |
+| 2 | bf1 | github:git.ouroath.com | null | 1 | screwdriver | sd@oath.com |
+| 3 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 1 | external | identity@oath.com |
+| 4 | identity | github:git.ouroath.com | identity_org1, identity_org2 | 0 | external | identity@oath.com |
+| 5 | iOS | github:git.ouroath.com | iOS_org1, iOS_org2 | 1 | external | ios@oath.com |
+
+### Cache server to store active *buildClusters* in memory when the service boot up. 
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url | Description
+| --- | --- | ---
+| `POST` | ` /buildClusters ` | ` { "name":"iOS", "scmContext":"github:git.ouroath.com", "scmOrganizations": "iOS_org1", "isActive":1, "managedBy": "screwdriver" } `
+| `GET` | `	/buildClusters ` | ` get list of buildClusters info `
+| `GET` | `	/buildClusters/:name ` | ` get a particular buildCluster info `
+| `DELETE` | ` /buildClusters/:name ` | ` delete buildCluster `
+
+
+### Cluster on-board
+
+1. Build cluster admin requesting access with cluster info and user credentials
+2. buildCluster table populated with cluster info
+3. SD admin to create queue based on #1
+4. SD admin authorize build cluster user and queue 
+
+
+## Flow
+### SD Validator
+
+	1. SD validator should validate if the annotated buildCluster in yaml is onboarded and active. 
+	
+### API to Redis
+
+	1. No change in existing implementation.
+
+### Master queue worker to Queue 
+
+	1. Master queue worker queries `buildClusters` cache for active records with cluster name from build info 
+	2. Validates if build job can be scheduled in specified buildCluster queue by validating scmContext + scmOrganization access and validate if blockedBy and other checks are success. 
```

Comment:
We shouldn't need to do the check any more, it should either be done by the validator or model when we tries to create the build. When it comes here, we can assume it has proper permission